### PR TITLE
Add a session backend to store session data in the database

### PR DESCRIPTION
**Pull request by blag**
This PR re-implements modified portions of [Flask-Session](https://github.com/fengsp/flask-session) to store session data in the database.

Things to note:

* This is my first major PR
* I'm not sure the added test file is in the right place
* Please double check my migration documentation is acceptable and complete

TODO:
* [x] Tests

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
What's the motivation for that change ?  It's quite a big change and you have not written why you want to implement it - this should be explained as part of the commit message - providing the context etc. Otherwise reviewer does not know what to review.

File: airflow/migrations/versions/c381b21cb7e4_add_session_table_to_db.py
Comment (ephraimbuddy) [inline airflow/migrations/versions/c381b21cb7e4_add_session_table_to_db.py]: We are trying to implement running migrations offline starting from 2.0.0 migration head and this would be a blocker.
PR: https://github.com/apache/airflow/pull/20962

Comment (blag) [reply airflow/migrations/versions/c381b21cb7e4_add_session_table_to_db.py]: Interesting, thanks for pointing me to that. I'll remove that then and have the upgrade assume that the table doesn't exist.

Comment (blag) [reply airflow/migrations/versions/c381b21cb7e4_add_session_table_to_db.py]: Marking this as resolved since I believe I fixed the issue. Please review again and let me know if I need to unresolve it.

File: airflow/models/__init__.py
Comment (ashb) [inline airflow/models/__init__.py]: Rather than doing this (and adding `airflow/models/session.py`) import it airflow.utils.db -- that's enough to get it loaded and reset and doing `reset db`

**Review by ashb — DISMISSED** [review]

**Comment by jedcunningham** [comment]
We need to add it to [NOTICE](https://github.com/apache/airflow/blob/main/NOTICE) as well.

File: airflow/_vendor/flask_session/sessions.py
Comment (jedcunningham) [inline airflow/_vendor/flask_session/sessions.py]:
```suggestion
import pickle
```

Let's keep it simple.

File: airflow/_vendor/flask_session/sessions.py
Comment (jedcunningham) [inline airflow/_vendor/flask_session/sessions.py]: In fact, can we serialize with json instead?

File: airflow/www/extensions/init_session.py
Comment (jedcunningham) [inline airflow/www/extensions/init_session.py]: We don't need a fallback - defaults are loaded automatically.

(You'd use fallbacks if it's possible for the config to not be in the default, e.g. if a provider could be installed on an earlier version of core airflow)

```suggestion
    selected_backend = conf.get('webserver', 'SESSION_BACKEND')
```

File: airflow/www/extensions/init_session.py
Comment (jedcunningham) [inline airflow/www/extensions/init_session.py]:
```suggestion
    elif selected_backend == 'database':
```

File: airflow/www/extensions/init_session.py
Comment (jedcunningham) [inline airflow/www/extensions/init_session.py:52]: We should probably raise if we get something else?

File: airflow/config_templates/config.yml
Comment (jedcunningham) [inline airflow/config_templates/config.yml]:
```suggestion
      version_added: 2.2.4
```

File: docs/apache-airflow/migrations-ref.rst
Comment (jedcunningham) [inline docs/apache-airflow/migrations-ref.rst]: This will need to be moved into the right spot for 2.2.4 (e.g. a different "revises id"), but I'm most likely going to pull another revision back to 2.2.4 tomorrow so hold off for now. At least this comment will remind us we need to do it.

File: airflow/_vendor/flask_session/db_models.py
Comment (jedcunningham) [inline airflow/_vendor/flask_session/db_models.py]: If we can do json, then we probably need to adjust this column type?

File: airflow/_vendor/flask_session/sessions.py
Comment (blag) [reply airflow/_vendor/flask_session/sessions.py]: Can we guarantee that all data being stored in the session is JSON-serializable?

**Comment by blag** [comment]
@jedcunningham Some of the code your wanted changes to is the vendored in code. I intentionally made as minimal modifications to that code as I needed, with the intent to possibly make a PR to upstream Flask-Session, get that merged, and migrate to using that (and removing our vendored code). The more and more in-depth modifications we make to the vendored code, the more difficult it will be to switch in the future.

I'm happy to make the changes you noted if you'd still like me to (please let me know), I just wanted to bring that to your attention first.

File: airflow/www/extensions/init_session.py
Comment (blag) [reply airflow/www/extensions/init_session.py:52]: I'm not sure how well this will fly, but I'd like to leave these undocumented - for now.

The upstream Flask-Session package supported these options, presumably because some users needed them. But every option we add is more branches and more code to possibly break and debug, so I'd like to keep the number of configurable options to a minimum.

So for these, I'd like to keep these available as escape hatches **for the interim** as we switch (most? all?) users over to database-backed sessions. If none of our users need these options then we can remove them and simply hard code the values. If we do have users who need these options, then we can explicitly document (and support) them later.

But if we really want to surface these options to users or if we want to hard code them out of the gate, I'm happy to make that change.

Comment (jedcunningham) [reply airflow/www/extensions/init_session.py:52]: I think we should:
- hard code the table name (otherwise we'd need to use it in the migration too, and honestly I don't think it's worth it)
- remove the `SESSION_KEY_PREFIX` (we wont ever share a backend)
- force `SESSION_USE_SIGNER` to True (we already require a secret_key)

**Comment by ashb** [comment]
> If we're trying to be able to easily understand what changes we've made to the vendored code, should we merge it in first with no changes? then make PRs on top of that which modify its behavior?

In a case or two previous to this we have merged the PR _without_ squashing (manually on the command line).

**Comment by blag** [comment]
@dstandish That's a good call, but it's looking less and less likely that our changes are going to be palatable to the maintainer of Flask-Session. If the changes we need won't ever be merged into Flask-Session, then our modified vendored code just becomes...our code, with an additional copyright line/header.

A "middle path" of this would be to start maintaining our own fork of Flask-Session, merge in the changes that we need, maybe merge in a few more unmerged PRs from Flask-Session, and have our fork become the friendlier (than Flask-Session) session package for Flask. But I don't want to sign the company up for that responsibility without some consensus and blessing, and I don't really want to maintain my own personal fork of Flask-Session specifically for Airflow.

I'm happy to do whatever it takes to get this merged, but I will need some guidance on how to get there.

File: airflow/_vendor/flask_session/sessions.py
Comment (jedcunningham) [inline airflow/_vendor/flask_session/sessions.py]: We might want to use `provide_session` instead?

**Pull request closed without merging**
